src/com/seafile/seadroid2/cameraupload/CameraSyncAdapter.java

```diff
@@ -0,0 +1,686 @@
+            String[] array = tempPath.split("/", 2);
+            String dir = array[0];
+            if (array.length == 2) {
+                tempPath = array[1];
+            } else {
+                tempPath = "";
+            }
+            if (dir.length() == 0) {
+                continue;
+            }
+
+            List<SeafDirent> dirs = dataManager.getDirentsFromServer(targetRepoId, base);
+            boolean found = false;
+            for (SeafDirent dirent : dirs) {
+                if (dirent.name.equals(dir))
+                    found = true;
+            }
+            if (!found)
+                dataManager.createNewDir(targetRepoId, base, dir);
+
+            base = Utils.pathJoin(base, dir);
+        }
+    }
+
+    @Override
+    public void onPerformSync(android.accounts.Account account,
+                              Bundle extras, String authority,
+                              ContentProviderClient provider,
+                              SyncResult syncResult) {
+
+        synchronized (this) {
+            cancelled = false;
+        }
+
+        Log.i(DEBUG_TAG, "Syncing images and video to " + account);
+
+        Log.d(DEBUG_TAG, "Camera upload target dir: "+settingsMgr.getCameraUploadDir());
+        Log.d(DEBUG_TAG, "Selected buckets for camera upload: "+settingsMgr.getCameraUploadBucketList());
+        Log.d(DEBUG_TAG, "is video upload allowed: "+settingsMgr.isVideosUploadAllowed());
+        Log.d(DEBUG_TAG, "is data plan allowed: "+settingsMgr.isDataPlanAllowed());
+
+        Log.d(DEBUG_TAG, "Media buckets available on this system: ");
+        for (GalleryBucketUtils.Bucket bucket: GalleryBucketUtils.getMediaBuckets(getContext())) {
+            Log.d(DEBUG_TAG, "Bucket id="+bucket.id+" name="+bucket.name);
+        }
+
+        // resync all media
+        if (extras.getBoolean(ContentResolver.SYNC_EXTRAS_INITIALIZE)) {
+            Log.i(DEBUG_TAG, "Doing a full resync");
+            dbHelper.cleanPhotoCache();
+        }
+
+        if (!settingsMgr.checkCameraUploadNetworkAvailable()) {
+            Log.d(DEBUG_TAG, "Not syncing because of data plan restriction.");
+            // treat dataPlan abort the same way as a network connection error
+            syncResult.stats.numIoExceptions++;
+            return;
+        }
+
+        Account seafileAccount = manager.getSeafileAccount(account);
+        DataManager dataManager = new DataManager(seafileAccount);
+
+        /**
+         * this should never occur, as camera upload is supposed to be disabled once the camera upload
+         * account signs out.
+         */
+        if (!seafileAccount.hasValidToken()) {
+            Log.d(DEBUG_TAG, "This account has no auth token. Disable camera upload.");
+            syncResult.stats.numAuthExceptions++;
+
+            // we're logged out on this account. disable camera upload.
+            ContentResolver.setIsSyncable(account, CameraUploadManager.AUTHORITY, 0);
```

Comment:
`ContentResolver.cancelSync(account.getAndroidAccount(), AUTHORITY);` cancels a pending sync request. But if the user is signed out, we should disable camera upload completely.